### ISSUE TITLE
cache is not invalidated for #include'd files

### ISSUE THREAD

**Issue by danielchasehooper** [ziglang/zig#11063]
### Zig Version

0.10.0-dev.871+bb05a8a08

### Steps to Reproduce

run this:
```
printf "#include \"mytypes.h\"
my_float_type public_func(my_float_type a) {
    return a + 1.0f;
}">main.c

printf "typedef float my_float_type;" > mytypes.h

zig build-lib  --enable-cache  main.c      #get everything cached

printf "" > mytypes.h                             #remove the definition of my_float_type

zig build-lib  --enable-cache  main.c      #this should fail, but doesn't
```

### Expected Behavior

the second `zig build-lib` should fail because `my_float_type` is no longer defined in `mytypes.h`

### Actual Behavior

zig doesn't doesn't recompile since it doesn't seem to notice that `my types.h` changed and `my_float_type` is no longer defined.

you can make the caching system pick up the change and (as expected) fail the build by modifying `main.c`:
```
printf "\n">>main.c
zig build-lib  --enable-cache  main.c 
```

changing `main.c` results in the expected compile error.

This is more serious when using `zig build` and you don't have the option to turn off the cache. This bug is serious enough  to completely prevent using c source with zig – your code changes don't make it into the binary.

**Comment by andrewrk** [comment]
Zig does include included files into the cache manifest. I'll try your steps to reproduce on Monday and if they indeed exhibit this bug then there is some much subtler problem happening than a missing feature of adding included files to the cache manifest.

**Comment by danielchasehooper** [comment]
@andrewrk Were you able to repro on your system? I’m happy to provide more info if helpful!

**Comment by andrewrk** [comment]
Yes, I was able to reproduce the issue. Note that when *not* passing `--enable-cache`, the global cache is still used for C files, and it behaves correctly - the header file makes it into the manifest. This appears to be an interaction with incremental compilation combined with `--enable-cache` (which is used by the build system) causing the misbehavior.

**Comment by danielchasehooper** [comment]
Great! If I recall correctly, I first noticed this issue in a project of mine using 'zig build'. So the issue may affect more mainstream use cases than just this redundant flag case.  I only used --enable-cache here to make the repro steps terse.

If it would be helpful, I can look into creating a repro case using zig build

**Comment by matu3ba** [comment]
user Moosichu on discord bisected this or a related behavior to
commit 67e31807df73ad2da992293fcdf1f6ea7ca67d2a and e3bed8d81dfd7198dd4c496f19a6791e27e41f26

**Comment by moosichu** [comment]
Yeah, those commits look like what Andrew suspected was the cause, but I've narrowed it to those two specifically - looking into it now.

**Comment by moosichu** [comment]
Going off:
> The two CacheMode values are `whole` and `incremental`.
> `incremental` is what we had before; `whole` is new.

My first test was to just disable whole caching in the zig compiler (forcing `cache_mode` to always be `.incremental`) - this does _not_ seem to be the same as what was there before, so something is at least a tad different. What I'm not sure of at this point is whether this is intentional (`incremental` could be changed in some ways because of `whole` existing, or if this is a bug). Will try and figure it out from looking at the code but will also ask Andrew for some help.

**Comment by moosichu** [comment]
Also:
> changing main.c results in the expected compile error.

In one project I've seen this (when I've tried to make an isolated test case), in my larger more complex project even modifying the "main" c source files results in changes not being picked up. This part is inconsistent.

**Comment by moosichu** [comment]
Without understanding the code that well yet, things don't seem quite right - I've been adding debug print statements to various parts of the cache system to try and understand what is/isn't happening. The next thing I'm going to do is do the same thing for the last known working version to see how those statements compare & how they differ, so I can more easily identify what is going wrong.

> In one project I've seen this (when I've tried to make an isolated test case), in my larger more complex project even modifying the "main" c source files results in changes not being picked up. This part is inconsistent.

A correction on this front - this does actually work but _only_ for the executable in my build.zig, the main source C files for my shared libraries don't seem to be getting checked in the cache however.

**Comment by moosichu** [comment]
I've put together a dummy PR #11365 as a place to test solutions to this problem.

I think the underlying problem with the system is that C headers just don't seem to be checked by the `whole` caching system at all (and also come kinds of C source files?) - do you how they are meant to make their way into the manifest? (my understanding of the whole system is still very patchy.)

I'm assuming doing so requires some kind of parsing of the C files at some point & adding the header paths in that case.

so in `Compilation.zig`, `fn addNonIncrementalStuffToCacheManifest(comp: *Compilation, man: *Cache.Manifest) !void` contains the lines:
```zig
    for (comp.c_object_table.keys()) |key| {
        _ = try man.addFile(key.src.src_path, null);
        man.hash.addListOfBytes(key.src.extra_flags);
    }
```
which just adds the C src file, what I'm assuming is that we need something akin to https://github.com/ziglang/zig/pull/10451/commits/e36718165cdc29b777392a3a343d92ccd1c6acf3 for c header includes, but I'm not sure if there is already a point at which these are processed for this purpose.

**Comment by matu3ba** [comment]
> need something akin to

I have no idea, how currently the system works, but overall I see 2 possible strategies to have a good solution without breaking code from caching.
In theory users could have different names for header and source files included together from other files with guards, so one must track/lookup when and where the respective symbol is available.

Strategy 1 (max perf)
1. requires to figure out how each .c file was combined.
3. requires to figure out how header files are combined by traversing from headers without dependencies to headers with dependencies to compare which parts of the dependency graph have changed
4. This whole thing needs to respect header include guards of the macro resolving, so to speed this up it could be (in the fast path) tracked where macro guards have been resolved to speed up the macro part.

Strategy 2 (simplicity)
If the goal is to do the naive strategy (not sure what C compilers use), then its sufficient to
1. track per file the dependencies
2. naively figure out which are the current leaf dependencies
3. apply the respective operations and compute with go to step 1

**Comment by moosichu** [comment]
So - the current caching system does add c include files to the manifests at some point (once it has committed to compiling c object files). But I've been struggling to locate how that happens & then add those to the "whole" cache.

**Comment by moosichu** [comment]
#11365 is now ready and working, ready for feedback on any stylistic stuff in there!

**Comment by andrewrk** [comment]
#11365  did not turn out to be a proper fix, apologies for closing the issue. See 06310e3d4eb47fed88b175891cb5865bb050f020 for more details.

**Comment by moosichu** [comment]
Anyone else is more than welcome to look at this as I don't have the bandwidth to do so at the moment myself, apologies for the lack of proper fix!

**Comment by moosichu** [comment]
I would say, currently using `zig build` for any C/C++ work is a no-go, so I don't know if it might be preferable to revert #10451 as well until this issue is fixed - or at least ensure that this issue is fixed before 0.10.0 ships.
